### PR TITLE
Doomslug

### PR DESCRIPTION
**Pull request by SkidanovAlex**
This PR consists of three commits: one reverts proof-of-stake-time, one completely removes the concept of weight from the codebase, and the final one contains the implementation of Doomslug. It might be easier to review the last commit without the first two, since the first two are very mechanical, but create enormous number of changes.

Since Github doesn't by default pick up the description of the multi-commit PR, pasting here the description of Doomslug commit:

Implementation and integration of Doomslug (see https://near.ai/doomslug).

Doomslug itself is in `chain/doomslug.rs`, and doesn't depend on chain or any
other code. It also always takes the current time as an argument. It
significantly simplifies testing.

Doomslug is stored on the client, not chain, because both block
production and approvals processing happens on client. Instantiating on
chain would require a slightly more complex interface (e.g. it would be
impossible to pass `me`).

Doomslug needs to know about the latest Tip. Instead of intercepting the
lowest level tip-updating routine (which is in storage), I update the
Tip in the client when we accept the new head. It could miss head
changes related to syncing and challenges. To be safe I always update
the head whenever I interact with Doomslug, so the head is guaranteed to
be accurate whenever we do anything related to Doomslug, but it might
get sent to Doomslug with a slight delay.

This change also solves a problem that approvals before were in a cache
hash map, therefore an adversary could have spammed a node with lots of
invalid approvals and remove all valid approvals from the cache,
resulting in a node not having enough approvals when producing
the block. New logic is more complex, see
`DoomslugApprovalsTrackersAtHeight` class.

Test plan
---------
1. Sanity tests (basic invariants for `set_tip` and approval processing)
2. A fuzzy test that tests safety and liveness under different times to
GST and deltas.
3. `test_catchup_sanity_blocks_produced_doomslug` verifies that heights
are properly skipped.
4. Also added checking doomslug invariant into `cross_shard_tx`, which
is known to evoke all kinds of weird structures (but `cross_shard_tx`
operates without the requirement to have 50% approvals on the block,
thus still causing lots of forkfullness).
5. A new version of `cross_shard_tx` that enables doomslug, but disables
tampering with the finality gadget. Thus the vanilla version tests heavy
forkfulness and tampers with FG, while the new doomslug version has
practically no forfulness due to doomslug, and doesn't stress the FG as
much, but does test block production with doomslug.

### CODE REVIEW

File: chain/chain/src/doomslug.rs [hunk 1]
```diff
@@ -0,0 +1,1004 @@
...
+/// Have that many iterations in the timer instead of `loop` to prevent potential bugs from blocking
+/// the node
+const MAX_TIMER_ITERS: usize = 20;
+
+/// How many heights ahead to track approvals. This needs to be sufficiently large so that we can
+/// recover after rather long network interruption, but not too large to consume too much memory if
+/// someone in the network spams with invalid approvals. Note that we will only store approvals for
+/// heights that are targeting us, which is once per as many heights as there are block producers,
+/// thus 10_000 heights in practice will mean on the order of one hundred entries.
+const MAX_HEIGHTS_AHEAD_TO_STORE_APPROVALS: BlockHeight = 10000;
```
Comment (ilblackdragon) [inline chain/chain/src/doomslug.rs]: nit: 10_000 (in comment you  did it but not here?)

File: chain/chain/tests/finality.rs [hunk 2]
```diff
@@ -165,7 +168,14 @@ fn create_block(
 }
 
 fn apr(account_id: AccountId, reference_hash: CryptoHash, parent_hash: CryptoHash) -> Approval {
```
Comment (ilblackdragon) [inline chain/chain/tests/finality.rs]: rename to `approve`?

File: chain/chain/src/doomslug.rs [hunk 3]
```diff
@@ -0,0 +1,1004 @@
...
+        target_height: BlockHeight,
+        is_endorsement: bool,
+    ) -> Option<Approval> {
+        self.me.as_ref().map(|me| {
+            Approval::new(
+                self.tip.block_hash,
+                self.tip.reference_hash,
+                target_height,
+                is_endorsement,
+                &**self.signer.as_ref().unwrap(),
```
Comment (ilblackdragon) [inline chain/chain/src/doomslug.rs]: this looks like something evil.
```
if let Some((me, signer)) = (self.me.as_ref(), self.signer.as_ref()) {
   ... 
    &*signer
}
```

also might be easier to pack `me` + `signer` into one object - BlockSigner.
We also need `BlockSigner` to get HSM signing going (because we will need to pass non-hashed data to HSM for signatures anyway ,so we need API that has specific `sign_block`, `sign_approval` APIs).
/\ this all sounds like a separate PR if you are not me :D

File: chain/client/src/client.rs [hunk 4]
```diff
@@ -97,15 +103,27 @@ impl Client {
         let num_block_producer_seats = config.num_block_producer_seats as usize;
         let data_parts = runtime_adapter.num_data_parts();
         let parity_parts = runtime_adapter.num_total_parts() - data_parts;
+
+        let doomslug = Doomslug::new(
+            block_producer.as_ref().map(|x| x.account_id.clone()),
```
Comment (ilblackdragon) [inline chain/client/src/client.rs]: may be let's move BlockProducer to primitives (and rename  to BlockSigner) and use it in Doomslug instead of splitting here?

File: chain/chain/src/doomslug.rs (hunk 3 above)
Comment (SkidanovAlex) [reply chain/chain/src/doomslug.rs]: I don't want to change anything related to HSM yet, it will be a separate effort.

Your approach still requires `&**signer`, because it's a reference to an `Arc`.

File: chain/chain/src/doomslug.rs [hunk 5]
```diff
@@ -0,0 +1,1004 @@
...
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub enum DoomslugThresholdMode {
+    NoApprovals,
+    HalfStake,
+}
+
+/// The result of processing an approval.
+///
+/// `None`
+/// `PassedThreshold(when)` - after processing this approval the block has passed the threshold set by
```
Comment (evgenykuzyakov) [inline chain/chain/src/doomslug.rs]: I'd suggest to move comments per item instead of having them as a top-level for enum.

File: chain/chain/src/doomslug.rs [hunk 6]
```diff
@@ -0,0 +1,1004 @@
...
+///                     `threshold_mode` (either one half of the total stake, or a single approval).
+///                     Once the threshold is hit, we want for `T(h - h_final) / 2` before producing
```
[the remark this answers is not in the document]
Comment (SkidanovAlex) [reply chain/chain/src/doomslug.rs]: The paper has been updated to reflect as to why we sleep

File: chain/chain/src/doomslug.rs [hunk 7]
```diff
@@ -0,0 +1,1004 @@
...
+            .iter()
+            .map(|approval| {
+                { account_id_to_stake.get(&approval.account_id) }.map_or(0, |stake| *stake)
+            })
+            .sum::<Balance>();
+
+        approved_stake > threshold
+    }
+
+    pub fn remove_witness(
```
[the remark this answers is not in the document]
Comment (SkidanovAlex) [reply chain/chain/src/doomslug.rs]: We remove it when we consume it to produce a block

File: chain/chain/src/doomslug.rs [hunk 8]
```diff
@@ -0,0 +1,1004 @@
...
+        last_ds_final_height: BlockHeight,
+    ) -> Option<Approval> {
+        self.tip = DoomslugTip { block_hash, reference_hash, height };
+
+        self.largest_ds_final_height = last_ds_final_height;
+        self.timer.height = height + 1;
+        self.timer.started = now;
+
+        self.approval_tracking
+            .retain(|h, _| *h > height && *h <= height + MAX_HEIGHTS_AHEAD_TO_STORE_APPROVALS);
```
[the remark this answers is not in the document]
Comment (SkidanovAlex) [reply chain/chain/src/doomslug.rs]: It's 10k iterations per each update of the head, which is acceptable.

Opened #2022  to track it.

File: chain/chain/src/doomslug.rs [hunk 9]
```diff
@@ -0,0 +1,1001 @@
...
+    /// `ReadyToProduce` if the block can be produced bypassing the timeout (i.e. has 2/3+ of
+    ///     approvals and 1/2+ of endorsements)
+    fn get_block_production_readiness(&mut self, now: Instant) -> DoomslugBlockProductionReadiness {
+        if self.approved_stake > self.total_stake * 2 / 3
+            && self.endorsed_stake > self.total_stake / 2
+        {
+            if self.time_passed_threshold == None {
+                self.time_passed_threshold = Some(now);
+            }
+            DoomslugBlockProductionReadiness::ReadyToProduce(self.time_passed_threshold.unwrap())
```
Comment (frol) [inline chain/chain/src/doomslug.rs]: I suggest to avoid `.unwrap()`, .e.g.:

```rust
let time_passed_threshold = if let Some(threshold) = self.time_passed_threshold {
    threshold
} else {
    self.time_passed_threshold = Some(now);
    now
};
DoomslugBlockProductionReadiness::ReadyToProduce(time_passed_threshold)
```

File: chain/chain/src/doomslug.rs [hunk 10]
```diff
@@ -0,0 +1,1001 @@
...
+            }
+        }
+
+        ret
+    }
+
+    pub fn create_approval(
+        &self,
+        target_height: BlockHeight,
+        is_endorsement: bool,
```
Comment (frol) [inline chain/chain/src/doomslug.rs]: I just want to bring this to our attention, that `bool`-based APIs are usually hard to follow when you read/maintain the code, e.g. the use looks like `self.create_approval(self.timer.height + 1, false)` (it is hard to tell what that `false` implies): http://blakesmith.me/2019/05/07/rust-patterns-enums-instead-of-booleans.html / https://www.reddit.com/r/programming/comments/ebxzp8/dont_use_booleans/

File: core/primitives/src/block.rs [hunk 11]
```diff
@@ -152,7 +148,9 @@ impl BlockHeaderInnerRest {
 #[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
 pub struct Approval {
     pub parent_hash: CryptoHash,
-    pub reference_hash: CryptoHash,
+    pub reference_hash: Option<CryptoHash>,
+    pub target_height: BlockHeight,
+    pub is_endorsement: bool,
```
Comment (frol) [inline core/primitives/src/block.rs]: If I understand the code correctly, `Approval` can be an enum, which would express the type in a safer way.

File: core/primitives/src/views.rs [hunk 12]
```diff
@@ -283,8 +280,9 @@ pub struct BlockHeaderView {
     pub challenges_result: ChallengesResult,
     pub last_quorum_pre_vote: CryptoHash,
     pub last_quorum_pre_commit: CryptoHash,
+    pub last_ds_final_block: CryptoHash,
     pub next_bp_hash: CryptoHash,
-    pub approvals: Vec<(AccountId, CryptoHash, CryptoHash, Signature)>,
+    pub approvals: Vec<(AccountId, CryptoHash, Option<CryptoHash>, BlockHeight, bool, Signature)>,
```
Comment (frol) [inline core/primitives/src/views.rs]: How do you feel about making this tuple a dedicated type (is it `Approval` structure?) where the arguments are named?

File: chain/client/src/client.rs [hunk 13]
```diff
@@ -666,6 +638,70 @@ impl Client {
...
+                last_ds_final_height,
+            );
+
+            if let Some(approval) = may_be_approval {
+                self.chain.process_approval(
+                    &self.block_producer.as_ref().map(|x| x.account_id.clone()),
+                    &approval,
+                )?;
+
+                self.collect_block_approval(&approval, true);
```
Comment (bowenwang1996) [inline chain/client/src/client.rs]: looks like this is not necessary

File: chain/chain/src/doomslug.rs [hunk 14]
```diff
@@ -0,0 +1,1001 @@
...
+            && self.endorsed_stake > self.total_stake / 2
+        {
+            if self.time_passed_threshold == None {
+                self.time_passed_threshold = Some(now);
+            }
+            DoomslugBlockProductionReadiness::ReadyToProduce(self.time_passed_threshold.unwrap())
+        } else if self.approved_stake > self.total_stake / 2
+            || self.threshold_mode == DoomslugThresholdMode::NoApprovals
+        {
+            if self.time_passed_threshold == None {
```
[the remark this answers is not in the document]
Comment (frol) [reply chain/chain/src/doomslug.rs]: Well, what I suggested above is to actually avoid this check and the unwrapping a few lines below in the first place.
